drivers/azure/azure.go

```diff
@@ -0,0 +1,464 @@
+	}
+	for _, f := range flags {
+		*f.target = fl.String(f.flag)
+		if *f.target == "" {
+			return requiredOptionError(f.flag)
+		}
+	}
+
+	// Optional flags or Flags of other types
+	d.Environment = fl.String(flAzureEnvironment)
+	d.OpenPorts = fl.StringSlice(flAzurePorts)
+	d.PrivateIPAddr = fl.String(flAzurePrivateIPAddr)
+	d.UsePrivateIP = fl.Bool(flAzureUsePrivateIP)
+	d.NoPublicIP = fl.Bool(flAzureNoPublicIP)
+	d.DockerPort = fl.Int(flAzureDockerPort)
+
+	// Set flags on the BaseDriver
+	d.BaseDriver.SSHPort = sshPort
+	d.SetSwarmConfigFromFlags(fl)
+
+	log.Debug("Set configuration from flags.")
+	return nil
+}
+
+// DriverName returns the name of the driver.
+func (d *Driver) DriverName() string { return driverName }
+
+// PreCreateCheck validates if driver values are valid to create the machine.
+func (d *Driver) PreCreateCheck() (err error) {
+	c, err := d.newAzureClient()
+	if err != nil {
+		return err
+	}
+
+	// Register used resource providers with current Azure subscription.
+	if err := c.RegisterResourceProviders(
+		"Microsoft.Compute",
+		"Microsoft.Network",
+		"Microsoft.Storage"); err != nil {
+		return err
+	}
+
+	// Validate if firewall rules can be read correctly
+	d.ctx.FirewallRules, err = d.getSecurityRules(d.OpenPorts)
+	if err != nil {
+		return err
+	}
+
+	// Check if virtual machine exists. An existing virtual machine cannot be updated.
+	log.Debug("Checking if Virtual Machine already exists.")
+	if exists, err := c.VirtualMachineExists(d.ResourceGroup, d.naming().VM()); err != nil {
+		return err
+	} else if exists {
+		return fmt.Errorf("Virtual Machine with name %s already exists", d.naming().VM())
+	}
+
+	// NOTE(ahmetalpbalkan) we could have done more checks here but Azure often
+	// returns meaningful error messages and it would be repeating the backend
+	// logic on the client side. Some examples:
+	//   - Deployment of a machine to an existing Virtual Network fails if
+	//     virtual network is in a different region.
+	//   - Changing IP Address space of a subnet would fail if there are machines
+	//     running in the Virtual Network.
+	log.Info("Completed machine pre-create checks.")
+	return nil
+}
+
+// Create creates the virtual machine.
+func (d *Driver) Create() error {
+	// NOTE(ahmetalpbalkan): We can probably parallelize the sh*t out of this.
+	// However that would lead to a concurrency logic and while creation of a
+	// resource fails, other ones would be kicked off, which could lead to a
+	// resource leak. This is slower but safer.
+	c, err := d.newAzureClient()
+	if err != nil {
+		return err
+	}
+
+	if err := c.CreateResourceGroup(d.ResourceGroup, d.Location); err != nil {
+		return err
+	}
+	if err := c.CreateAvailabilitySetIfNotExists(d.ctx, d.ResourceGroup, d.AvailabilitySet, d.Location); err != nil {
+		return err
+	}
+	if err := c.CreateNetworkSecurityGroup(d.ctx, d.ResourceGroup, d.naming().NSG(), d.Location, d.ctx.FirewallRules); err != nil {
+		return err
```

Comment:
See above comment. I am trying to understand, if the objects created earlier need to be removed in the case of failures.